.github/workflows/prepare-bundled-themes.yml

```diff
@@ -0,0 +1,157 @@
+  # or the commit hash for any other events.
+  group: ${{ github.workflow }}-${{ github.event_name == 'pull_request' && github.head_ref || github.sha }}
+  cancel-in-progress: true
+
+jobs:
+  # Prepares bundled themes for release.
+  #
+  # Performs the following steps:
+  # - Checks out the repository.
+  # - Installs Node.js (only when theme has a build process)
+  # - Installs npm dependencies (only when theme has a build process)
+  # - Runs the theme build script (only when theme has a build process)
+  # - Ensures version-controlled files are not modified or deleted.
+  # - Creates a ZIP file of the theme for wordpress.org.
+  # - Uploads the ZIP file as a workflow artifact.
+  bundle-theme:
+    name: Prepare ${{ matrix.theme }}
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    if: ${{ github.repository == 'WordPress/wordpress-develop' || github.event_name == 'pull_request' }}
+    strategy:
+      fail-fast: false
+      matrix:
+        theme: [
+            'twentytwentythree',
+            'twentytwentytwo',
+            'twentytwentyone',
+            'twentytwenty',
+            'twentynineteen',
+            'twentyseventeen',
+            'twentysixteen',
+            'twentyfifteen',
+            'twentyfourteen',
+            'twentythirteen',
+            'twentytwelve',
+            'twentyeleven',
+            'twentyten'
+        ]
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8 # v3.1.0
+        with:
+          ref: ${{ github.event_name == 'workflow_dispatch' && inputs.branch || github.ref }}
+
+      - name: Install Node.js
+        uses: actions/setup-node@8c91899e586c5b171469028077307d293428b516 # v3.5.1
+        if: ${{ contains( fromJson( '["twentynineteen", "twentytwenty", "twentytwentyone"]' ), matrix.theme ) }}
+        with:
+          node-version-file: '.nvmrc'
+          cache: npm
+          cache-dependency-path: src/wp-content/themes/${{ matrix.theme }}/package-lock.json
+
+      - name: Install npm dependencies
+        if: ${{ contains( fromJson( '["twentynineteen", "twentytwenty", "twentytwentyone"]' ), matrix.theme ) }}
+        run: npm ci
+        working-directory: src/wp-content/themes/${{ matrix.theme }}
+
+      - name: Build theme
+        if: ${{ contains( fromJson( '["twentynineteen", "twentytwenty", "twentytwentyone"]' ), matrix.theme ) }}
+        run: npm run build
+        working-directory: src/wp-content/themes/${{ matrix.theme }}
+
+      - name: Ensure version-controlled files are not modified or deleted
+        if: ${{ contains( fromJson( '["twentynineteen", "twentytwenty", "twentytwentyone"]' ), matrix.theme ) }}
+        run: git diff --exit-code
+        working-directory: src/wp-content/themes/${{ matrix.theme }}
+
+      - name: Create theme ZIP file
+        run: zip -r ${{ matrix.theme }}.zip src/wp-content/themes/${{ matrix.theme }}
```

Comment:
This will need to exclude a few things:

* node_modules, I think they're all dev dependencies.
* dot files
* composer, npm config files

`src/wp-content/themes/${{ matrix.theme }}` is included in the zip file, ie `style.css` is four directories down. You probably want to make `themes` the working directory.

Reply:
I've pushed an update to exclude `node_modules` directories.

I think in the past we have always included `package.json`/`package-lock.json` files so that anyone could download the theme and get up and running building with it.

It seems only TT1 has dotfiles. I couldn't find any specific discussion around `.stylelint*` related files that start with `.`s. It seems that those were included in versions [1.4](https://themes.svn.wordpress.org/twentytwentyone/1.4/) [and](https://themes.svn.wordpress.org/twentytwentyone/1.3/) [earlier](https://themes.svn.wordpress.org/twentytwentyone/1.2/), but not in 1.5 or 1.6.

Since these files configure how build scripts run, I'd consider that a regression and we should probably reintroduce those.

As for changing the working directory, I don't feel strongly here. I decided to use the main working directory of the runner to just place the ZIP in the top directory.